jormungandr/src/leadership/process.rs

```diff
@@ -0,0 +1,630 @@
+                            .to_genesis_praos_builder()
+                            .unwrap()
+                            .set_consensus_data(&node_id, &vrfproof.into());
+                        Either::B(Either::B(
+                            enclave
+                                .query_header_genesis_praos_finalize(final_builder, event.id)
+                                .map(|h| Block {
+                                    header: h.generalize(),
+                                    contents,
+                                })
+                                .map_err(|_| {
+                                    // the only reason this may happen is if the leader
+                                    // has been removed from the enclave
+                                    // which is not supported at the moment
+                                    unimplemented!()
+                                }),
+                        ))
+                    }
+                }
+            });
+
+            let event_logs_success = event_logs.clone();
+            let send_block = signing.and_then(|block| {
+                let id = block.header.hash();
+                let chain_length: u32 = block.header.chain_length().into();
+                sender
+                    .send(BlockMsg::LeadershipBlock(block))
+                    .map_err(|_send_error| LeadershipError::CannotSendLeadershipBlock)
+                    .and_then(move |_| {
+                        event_logs_success
+                            .set_status(LeadershipLogStatus::Block {
+                                block: id.into(),
+                                chain_length,
+                            })
+                            .map_err(|()| unreachable!())
+                    })
+            });
+
+            let timeout = Timeout::new_at(send_block, deadline)
+                .or_else(move |timeout_error| {
+                    error!(logger, "Eek... took too long to process the event..." ; "reason" => %timeout_error);
+                    event_logs.set_status(LeadershipLogStatus::Rejected {
+                        reason: "Failed to compute the schedule within time boundaries".to_owned()
+                    }).map_err(|()| unreachable!())
+                })
+                .map(|_| self);
+
+            Either::A(timeout)
+        } else {
+            // the event happened out of bounds, ignore it and move to the next one
+            error!(
+                logger,
+                "Eek... we missed an event schedule, system time might be off?"
+            );
+
+            let tell_user_about_failure = event_logs
+                .set_status(LeadershipLogStatus::Rejected {
+                    reason: "Missed the deadline to compute the schedule".to_owned(),
+                })
+                .map_err(|()| unreachable!());
+
+            Either::B(Either::B(tell_user_about_failure.map(|()| self)))
```

Comment:
This could be broken up into smaller functions so that there are no nested `Either`s.

Reply:
yes and trying to remove the unwraps in this function too.